utilities/include/par.h

```diff
@@ -0,0 +1,156 @@
+#pragma once
+#include <thrust/binary_search.h>
+#include <thrust/gather.h>
+#include <thrust/logical.h>
+#include <thrust/remove.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+#include <thrust/uninitialized_copy.h>
+#include <thrust/unique.h>
+
+#include <thrust/system/cpp/execution_policy.h>
+#include <thrust/system/omp/execution_policy.h>
+
+#if THRUST_DEVICE_SYSTEM == THRUST_DEVICE_SYSTEM_CUDA
+#include <thrust/system/cuda/execution_policy.h>
+#endif
+
+namespace manifold {
+
+void check_cuda_available();
+#if THRUST_DEVICE_SYSTEM == THRUST_DEVICE_SYSTEM_CUDA
+extern int CUDA_ENABLED;
+#else
+constexpr int CUDA_ENABLED = 0;
+#endif
+
+enum class ExecutionPolicy {
+  ParUnseq,
+  Par,
+  Seq,
+};
+
+constexpr ExecutionPolicy ParUnseq = ExecutionPolicy::ParUnseq;
+constexpr ExecutionPolicy Par = ExecutionPolicy::Par;
+constexpr ExecutionPolicy Seq = ExecutionPolicy::Seq;
+
+// ExecutionPolicy:
+// - Sequential for small workload,
+// - Parallel (CPU) for medium workload,
+// - GPU for large workload if available.
+inline ExecutionPolicy autoPolicy(int size) {
+  // some random numbers
+  if (size <= (1 << 12)) {
+    return Seq;
+  }
+  if (size <= (1 << 16) || CUDA_ENABLED != 1) {
+    return Par;
+  }
+  return ParUnseq;
+}
+
+#if THRUST_DEVICE_SYSTEM == THRUST_DEVICE_SYSTEM_CUDA
+#define THRUST_DYNAMIC_BACKEND_VOID(NAME, SUFFIX)                              \
+  template <typename... Args>                                                  \
+  void NAME##SUFFIX(ExecutionPolicy policy, Args... args) {                    \
+    switch (policy) {                                                          \
+    case ExecutionPolicy::ParUnseq:                                            \
+      thrust::NAME(thrust::cuda::par, args...);                                \
+      break;                                                                   \
+    case ExecutionPolicy::Par:                                                 \
+      thrust::NAME(thrust::omp::par, args...);                                 \
```

Comment:
So, if you wanted TBB, that would replace OMP here, right? Maybe for build rules we need something like `ParUnseq = CUDA | NONE` and `Par = OMP | TBB | NONE`? I guess `NONE` is mostly about building for compilers that don't support these.

Reply:
Yes

Reply:
I guess I will update this and the CMake script after #125 is merged.